File: leetcode/python/p3402.py
from typing import List


class Solution:
    def isPrefixAndSuffix(self, str1, str2):
        return str2.startswith(str1) and str2.endswith(str1)

    def countPrefixSuffixPairs(self, words: List[str]) -> int:
        n = len(words)
        ans = 0
        for i in range(n):
            for j in range(i+1, n):
                if self.isPrefixAndSuffix(words[i], words[j]):
                    ans += 1
        return ans


if __name__ == "__main__":
    solution = Solution()
    print(solution.countPrefixSuffixPairs(["a", "aba", "ababa", "aa"]))

File: leetcode/3000/p3402.cpp


class Solution {
   public:
    bool isPrefixAndSuffix(const std::string& str1, const std::string& str2) {
        if (str1.length() > str2.length()) return false;
        return str2.substr(0, str1.length()) == str1 &&
               str2.substr(str2.length() - str1.length()) == str1;
    }

    int countPrefixSuffixPairs(std::vector<std::string>& words) {
        int n = words.size();
        int ans = 0;
        for (int i = 0; i < n; i++) {
            for (int j = i + 1; j < n; j++) {
                if (isPrefixAndSuffix(words[i], words[j])) {
                    ans++;
                }
            }
        }
        return ans;
    }
};
